### PR TITLE
add ``Trainer`` compatibility with ``Denoiser``

### PR DESCRIPTION
**Pull request by Tmodrzyk**
This PR follows the suggestion from @tachella in issue #593. 
When the model passed to the ``Trainer`` is a ``Denoiser``, it is wrapped in an ``ArtifactRemoval`` class. It is unwrapped at the end of the training, and during intermediate checkpoints to preserve the ``state_dict``.

:warning:  Note that this fixes the issue for the specific case of ``Denoiser``, but if we later want to support more general models (e.g., those conditioned on additional inputs such as segmentations or other images), the Trainer may require further generalization.
Let me know if I should add a gallery example or notebook on how to train a denoiser, I think that is a fairly common use case of the library. :smile: 

### Checks to be done before submitting your PR

- [x] `python3 -m pytest deepinv/tests` runs successfully.
- [x] `black .` runs successfully.
- [ ] `make html` runs successfully (in the `docs/` directory).
- [ ] Updated docstrings related to the changes (as applicable).
- [ ] Added an entry to the [CHANGELOG.rst](../CHANGELOG.rst).

### CODE REVIEW

**Comment by Tmodrzyk** [comment]
It seems that a lot of models / class inherit from ``Denoiser`` without ever using the noise level ``sigma``, for instance, `DnCNN`, `AutoEncoder`, `Restormer`, `SwinIR`, `UNet`, etc.
In this PR, I check whether a model is an instance of `Denoiser` before wrapping it in `ArtifactRemoval`, to ensure that `sigma` is passed instead of `physics`. 
But this test does not work as intended if some denoisers don't use ``sigma``.

I'm wondering if we really want these models to inherit from `Denoiser` if they don’t use the noise level? Wouldn't it be cleaner for them to simply subclass `nn.Module` unless they explicitly require `sigma`?

File: deepinv/training/trainer.py [hunk 1]
```diff
@@ -233,6 +234,12 @@ class Trainer:
     verbose_individual_losses: bool = True
     show_progress_bar: bool = True
 
+    def setup_denoiser(self):
```
Comment (Andrewwango) [inline deepinv/training/trainer.py:243]: just a very short one line docstring here please

File: deepinv/training/trainer.py [hunk 2]
```diff
@@ -15,6 +15,7 @@
 from deepinv.utils.plotting import prepare_images
 from torchvision.utils import save_image
 import inspect
+from deepinv.models import ArtifactRemoval, Denoiser
 
 
 @dataclass
```
Comment (Andrewwango) [inline deepinv/training/trainer.py:28]: Please can you add a small note to the docstring about `model` that if you pass in a denoiser, it automatically gets converted to an `ArtifactRemoval` for training?

**Review by Andrewwango — CHANGES_REQUESTED** [review]
Hey Thibaut thanks for the PR!

To fix the tests and docs please can you apply the following hotfixes (these are planned to be fixed in #409 ):

1. Add `model(y, physics)` here:
https://github.com/deepinv/deepinv/blob/464ec13d13c0bbc6dfd0b7f133ea9b6a161d1636/deepinv/loss/adversarial/uair.py#L77
2. Wrap this unet in a `ArtifactRemoval(UNet(...), mode="direct"` (the PR should work without this fix, but it's good to be explicit about best practice in the docs):
https://github.com/deepinv/deepinv/blob/464ec13d13c0bbc6dfd0b7f133ea9b6a161d1636/examples/adversarial-learning/demo_gan_imaging.py#L112
3. Likewise, wrap the Unet here:
https://github.com/deepinv/deepinv/blob/464ec13d13c0bbc6dfd0b7f133ea9b6a161d1636/deepinv/tests/test_adversarial.py#L19

**Comment by Andrewwango** [comment]
Additional comments:

> Let me know if I should add a gallery example or notebook on how to train a denoiser, I think that is a fairly common use case of the library. 😄

That would be super cool! I think the "Basics" section could really benefit from this example

> I'm wondering if we really want these models to inherit from Denoiser if they don’t use the noise level? Wouldn't it be cleaner for them to simply subclass nn.Module unless they explicitly require sigma?`

I think it's ok as is, as long as we are clear in the docs that these models you listed are by default Denoisers, which you can convert into a Reconstructor if you want.

File: deepinv/training/trainer.py (hunk 2 above)
Comment (Tmodrzyk) [reply deepinv/training/trainer.py:28]: Done in [0844921](https://github.com/deepinv/deepinv/pull/614/commits/0844921e8d8f788a8ee347bbc5fa9743c617b363)

**Comment by Tmodrzyk** [comment]
Hi @Andrewwango, I think this is good to go ? Let me know if you think the example is enough :)

File: examples/basics/demo_train_denoiser.py [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+#       The number of epochs and the batch size should be based on the available hardware and the size of the dataset.
+#
+
+verbose = True  # print training information
+wandb_vis = False  # plot curves and images in Weight&Bias
+
+loss = dinv.loss.SupLoss(metric=dinv.metric.MSE())
+metrics = [dinv.metric.PSNR(), dinv.metric.SSIM()]
+
+optimizer = torch.optim.Adam(
```
Comment (Andrewwango) [inline examples/basics/demo_train_denoiser.py:110]: How does this work? After the model gets wrapped, surely optimizer is tracking the wrong parameters?

File: examples/basics/demo_train_denoiser.py [hunk 4]
```diff
@@ -0,0 +1,144 @@
+r"""
+Training a denoiser.
+====================================================================================================
+
+This example shows how to train a standard denoiser such as the [DRUNet by Zhang et al., (2020)](https://arxiv.org/abs/2008.13751)! using the :class:`deepinv.Trainer` class.
```
Comment (Andrewwango) [inline examples/basics/demo_train_denoiser.py:5]: Ref bib instead following #575

File: examples/basics/demo_train_denoiser.py [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+CKPT_DIR = BASE_DIR / "ckpts"
+
+# Set the global random seed from pytorch to ensure reproducibility of the example.
+torch.manual_seed(0)
+device = dinv.utils.get_freer_gpu() if torch.cuda.is_available() else "cpu"
+
+# %%
+# Load train and eval datasets
+# --------------------------------------------------------------------------------------------
+# In this example, we use the CBSD68 dataset for training and the set3c dataset for validation.
```
Comment (Andrewwango) [inline examples/basics/demo_train_denoiser.py:33]: Ref dinv classes for the datasets

File: examples/basics/demo_train_denoiser.py [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+    ckp_interval=2,
+    save_path=CKPT_DIR,
+)
+
+trainer.train()
+
+model = trainer.load_best_model()
+
+# %%
+# Test the network
```
Comment (Andrewwango) [inline examples/basics/demo_train_denoiser.py:138]: Can we plot some images please

File: examples/basics/demo_train_denoiser.py [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+# --------------------------------------------
+# We can now test the trained network using the :func:`deepinv.test` function.
+#
+# Here the testing function will compute the PSNR and SSIM metrics on the evaluation dataset.
+
+trainer.test(eval_dataloader)
```
Comment (Andrewwango) [inline examples/basics/demo_train_denoiser.py:144]:
```
Test results:
PSNR no learning: 22.182 +- 14.746
PSNR: 15.195 +- 1.237
SSIM no learning: 0.526 +- 0.334
SSIM: 0.606 +- 0.029
```
The results aren't great - maybe should we train starting from a pretrained model? Or should we just test on the train set?

File: examples/basics/demo_train_denoiser.py (hunk 3 above)
Comment (Tmodrzyk) [reply examples/basics/demo_train_denoiser.py:110]: From my understanding since ``ArtifactRemoval`` doesn't add any trainable parameters, the optimizer still tracks the same parameters ? 
When I check in the demo file ``model.named_parameters()`` and ``trainer.model.named_parameters()`` are the same.

Comment (Tmodrzyk) [reply examples/basics/demo_train_denoiser.py:110]: Ok so it seems that the model is not learning right now so you were probably right about the optimizer not tracking the right parameters :smiling_face_with_tear:
